src/drivers/intel/dmic.c

```diff
@@ -1563,17 +1563,17 @@ static void dmic_irq_handler(void *data)
 
 	if (val0 & OUTSTAT0_ROR_BIT) {
 		dai_err(dai, "dmic_irq_handler(): full fifo A or PDM overrun");
-		dai_write(dai, OUTSTAT0, val0);
 		dmic_stop(dai, dmic->state == COMP_STATE_ACTIVE);
 		dmic->state = COMP_STATE_PREPARE;
 	}
+	dai_write(dai, OUTSTAT0, val0);
```

Comment:
this changes the sequence for error handling, should you move the two writes before the tests and factor the two dmic_stop?

e.g.

````c
needs_stop = false;
val0 = dai_read(dai, OUTSTAT0);
val1 = dai_read(dai, OUTSTAT1);

dai_write(dai, OUTSTAT0, val0);
dai_write(dai, OUTSTAT1, val1);

if (val0 & OUTSTAT0_ROR_BIT) {
    dai_err(dai, "dmic_irq_handler(): full fifo A or PDM overrun");
    needs_stop = true;
}

if (val1 & OUTSTAT1_ROR_BIT) {
    dai_err(dai, "dmic_irq_handler(): full fifo B or PDM overrun");
    needs_stop = true;
}

if (needs_stop) {
    dmic_stop(dai, dmic->state == COMP_STATE_ACTIVE);
    dmic->state = COMP_STATE_PREPARE;
}
````

Reply:
Ack, thanks @plbossart . I'll do a non-draft cleanup version. I'm also not 100% sure calling dmic_stop() is safe here. To me it seems possible that a dmic interrupt is raised while a DSP core is already running dmic_stop() (for normal pipeline stop).

I think the primary reason to stop the dmic here (rather than relying on pipeline xrun handling) is to avoid spamming the trace log when a xrun happens.